ecosystem/typescript/sdk/src/transaction_builder/builder_utils.ts

```diff
@@ -72,84 +72,108 @@ export function ensureBigInt(val: number | bigint | string): bigint {
-      addr = argVal;
-    } else {
-      throw new Error("Invalid account address.");
+}
+
+function serializeAddress(argVal: any, serializer: Serializer) {
+  let addr: AccountAddress;
+  if (typeof argVal === "string" || argVal instanceof HexString) {
+    addr = AccountAddress.fromHex(argVal);
+  } else if (argVal instanceof AccountAddress) {
+    addr = argVal;
+  } else {
+    throw new Error("Invalid account address.");
+  }
+  addr.serialize(serializer);
+}
+
+function serializeVector(argVal: any, argType: TypeTagVector, serializer: Serializer, depth: number) {
+  // We are serializing a vector<u8>
+  if (argType.value instanceof TypeTagU8) {
+    if (argVal instanceof Uint8Array) {
+      serializer.serializeBytes(argVal);
+      return;
     }
-    addr.serialize(serializer);
-    return;
-  }
-  if (argType instanceof TypeTagVector) {
-    // We are serializing a vector<u8>
-    if (argType.value instanceof TypeTagU8) {
-      if (argVal instanceof Uint8Array) {
-        serializer.serializeBytes(argVal);
-        return;
-      }
-
-      if (typeof argVal === "string") {
-        serializer.serializeStr(argVal);
-        return;
-      }
+    if (argVal instanceof HexString) {
+      serializer.serializeBytes(argVal.toUint8Array());
+      return;
     }
-
-    if (!Array.isArray(argVal)) {
-      throw new Error("Invalid vector args.");
+    if (typeof argVal === "string") {
+      serializer.serializeStr(argVal);
+      return;
     }
+    // If it isn't any of those types, then it must just be an actual array of numbers
+  }
 
-    serializer.serializeU32AsUleb128(argVal.length);
-
-    argVal.forEach((arg) => serializeArg(arg, argType.value, serializer));
-    return;
+  if (!Array.isArray(argVal)) {
+    throw new Error("Invalid vector args.");
   }
 
-  if (argType instanceof TypeTagStruct) {
-    const { address, module_name: moduleName, name } = (argType as TypeTagStruct).value;
-    if (
-      `${HexString.fromUint8Array(address.address).toShortString()}::${moduleName.value}::${name.value}` !==
-      "0x1::string::String"
-    ) {
-      throw new Error("The only supported struct arg is of type 0x1::string::String");
-    }
-    assertType(argVal, ["string"]);
+  serializer.serializeU32AsUleb128(argVal.length);
 
+  argVal.forEach((arg) => serializeArgInner(arg, argType.value, serializer, depth + 1));
+}
+
+function serializeStruct(argVal: any, argType: TypeTag, serializer: Serializer, depth: number) {
+  const { address, module_name: moduleName, name, type_args: typeArgs } = (argType as TypeTagStruct).value;
+  const structType = `${HexString.fromUint8Array(address.address).toShortString()}::${moduleName.value}::${name.value}`;
+  if (structType === "0x1::string::String") {
+    assertType(argVal, ["string"]);
     serializer.serializeStr(argVal);
-    return;
+  } else if (structType === "0x1::object::Object") {
+    serializeAddress(argVal, serializer);
+  } else if (structType === "0x1::option::Option") {
```

Comment:
how does `0x1::option::Option` represented in the ABI? can we add tests to show it is parsed as expected here https://github.com/aptos-labs/aptos-core/blob/main/ecosystem/typescript/sdk/src/tests/unit/type_tag.test.ts?